### PR TITLE
Enemy Pathfinding

### PR DESCRIPTION
**Pull request by ChaseBENNC9**
## Describe your changes

Enemy Pathfinding, Movement, player detection

# Issue

#2 

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by ChaseBENNC9** [comment]
https://docs.unity3d.com/ScriptReference/AI.NavMeshAgent.html

**Comment by ChaseBENNC9** [comment]
Ready for testing, the enemy should wander around the plane randomly. Currently the player detection is not working so its just the movement for now, I wan't to change the movement to be smoother but would like some feedback to see how it feels at the moment Ideally when the player detection is added the enemy would move to a position and check for any players close to it before either throwing a snowball or moving to a new spot.

**Review by PBWiseman — COMMENTED** [review]
The movement and pathfinding looks really good. I mentioned in our discussion on teams some ideas for implementing the player detection. The turning is still a little janky. Do we want sudden 8 way turning or smooth turning because currently there is both? And the smooth turning parts are sometimes slow enough that the target is reached and a new turn started before the first one is finished.

**Comment by ChaseBENNC9** [comment]
> The movement and pathfinding looks really good. I mentioned in our discussion on teams some ideas for implementing the player detection. The turning is still a little janky.

 Do we want sudden 8 way turning or smooth turning because currently there is both? And the smooth turning parts are sometimes slow enough that the target is reached and a new turn started before the first one is finished.
I think sudden turning and I am not sure why sometimes it rotates slowly. But I will work on that and start implementing the other ideas for player detection

**Comment by PBWiseman** [comment]
> Do we want sudden 8 way turning or smooth turning because currently there is both? And the smooth turning parts are sometimes slow enough that the target is reached and a new turn started before the first one is finished. I think sudden turning and I am not sure why sometimes it rotates slowly. But I will work on that and start implementing the other ideas for player detection

There is some code in my branch for turning the player. I just set transform.forward to the movement direction. Might help?

**Comment by ChaseBENNC9** [comment]
The enemy should now only face in one of the 8 directions. at this point it is sudden turning and not smoothly rotating between each point. Player detection should also work in which it shows a blue line towards a player sphere if it is within range, and this is also with a slight buffer, so it doesn't have to always be exact.

File: Assets/Scripts/EnemyMovement.cs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        //             0,
+        //             GetClosestPlayer().transform.position.z
+        //         ) - new Vector3(transform.position.x, 0, transform.position.z)
+        //     ).normalized; //Sets the forward direction of the agent to the direction to the player
+        //     state = EnemyStates.TargetingPlayer;
+        // }
+
+
+        // else
+        {
```
Comment (PBWiseman) [inline Assets/Scripts/EnemyMovement.cs:64]: There's an extra open and close bracket here, I assume from the commented else

File: Assets/Scripts/EnemyMovement.cs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        foreach (GameObject player in players) //Loops through all the players
+        {
+            float distance = Vector3.Distance(transform.position, player.transform.position); //Gets the distance between the agent and the player
+            if (distance < closestDistance) //Checks if the distance is less than the closest distance
+            {
+                closestDistance = distance; //Sets the closest distance to the distance
+                closestPlayer = player; //Sets the closest player to the player
+            }
+        }
+        state = EnemyStates.Moving; //Sets the state to idle
```
Comment (PBWiseman) [inline Assets/Scripts/EnemyMovement.cs:142]: The comment says idle state and the code says moving. I assume just an old comment?

File: Assets/Scripts/EnemyMovement.cs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    TargetingPlayer,
+    ThrowingSnowball
+}
+
+public class EnemyMovement : MonoBehaviour
+{
+    private NavMeshAgent agent; //The NavMeshAgent component
+    public NavMeshSurface surface; //The NavMeshSurface component
+    private Vector3 randomPosition; //The random position to move to
+    public GameObject sphere; //The sphere prefab
```
Comment (PBWiseman) [inline Assets/Scripts/EnemyMovement.cs:23]: This doesn't seem to be used anymore, even for testing?

**Review by PBWiseman — CHANGES_REQUESTED** [review]
Impressive work! The tracking and movement work really well. I assume the snowball throwing will be another else if state statement in the update? The only issue I can see at the moment is cleaning up the leftover debug statements and blocks of commented code that you aren't saving for later. Plus a few other nitpicks I left comments for on the lines of code

**Comment by ChaseBENNC9** [comment]
@PBWiseman All issues should be resolved

**Review by PBWiseman — CHANGES_REQUESTED** [review]
Everything looks good except for the fact that it seems to be trying to run at 600 fps? We may want to cap that. My computer is not happy with this
16b02a82-321b-4bfc-bf31-2d51f86f9c52

**Review by erokasyuart — CHANGES_REQUESTED** [review]
Good commenting, nice seeing it all being explained, especially when getting lost in all those if statements!

One issue I have is this is all that happens in the scene?
image

I checked and I'm up to date on the branch, have I just come in at the wrong time? The enemy doesn't move, just makes a line.

**Review by PBWiseman — APPROVED** [review]

**Review by erokasyuart — APPROVED** [review]

**Pull request merged**